### ISSUE TITLE
On APIs for 0.20.x, create methods wrongly indicate that they return objects

### ISSUE THREAD

**Issue by GregOriol** [killbill/killbill-client-php#40]
On APIs for 0.20.x, generated by swagger, create* methods like createAccount are wrongly documented as returning an object. In fact, they don't: the POST request returns a 201 code and a Location header to follow to get the object, not the object itself. The generated code doesn't take that Location header into account and tries to deserialize the object from the response itself, and thus returns null.

So far, there doesn't seem to be a good/quick solution to this. The code using the client must not take that part of the documentation into account, must consider that the create* methods return nothing, and must get the created object using another request.

**Comment by pierre** [comment]
IIRC this is/was a Swagger limitation with our model of returning 201.

Other clients _follow_ the redirect by default unless specified in their respective config (in which case, the method returns `null`):

* https://github.com/killbill/killbill-client-ruby/blob/03726f75db8704d21829ec80f5c57d2bcff5681d/lib/killbill_client/models/account.rb#L79
* https://github.com/killbill/killbill-client-java/blob/74aa69ec74c5cd1674d0973e631b463cf5ba953a/src/main/java/org/killbill/billing/client/KillBillHttpClient.java#L344-L356
* https://github.com/killbill/kbcli/blob/ae87ff915e3e49baf19e5b82d01946d7086d9697/kbclient/account/account_client.go#L481-L503

**Comment by GregOriol** [comment]
I guess we'll have to modify swagger codegen templates to handle this

**Comment by pierre** [comment]
Yup, you probably want to do it anyways to introduce some form of context (_RequestOptions_ in other clients) to abstract the headers values (tenant credentials, audit, etc.).

**Comment by GregOriol** [comment]
Do you use custom templates to generate the java client for example? I can't find any indication about it in the repository

**Comment by pierre** [comment]
We do, but they aren't open-source unfortunately.

The Go ones are here though: https://github.com/killbill/go-swagger

**Comment by GregOriol** [comment]
Ok ok
The go ones don't help that much as they are not using swagger-codegen nor mustache/handlebars templates

**Comment by pierre** [comment]
Oh, right. Do you want to see anything specifically? Happy to share some snippets.

The repo isn't open-source for historical reasons (we'd just need to clean things up to make it public, but it's pretty low on our TODO list... 😅).

**Comment by GregOriol** [comment]
I fixed this "issue" in PR #39 by refactoring the templates and adding the 201+location handling

I'd be happy to see how you handled things like the AuditLevel type because currently in php it is just a string, could be better

**Comment by pierre** [comment]
We have a similar issue for the Python client unfortunately. For Go, we have custom code to handle enums.

In Java, it's easier because we can import the `killbill-api` JAR directly (we have extended the Java generator with vendor extensions to support it).

**Comment by GregOriol** [comment]
I guess we could generate some php class/constants for it, but I can't find how to type it using Swagger: in the swagger.json, the type is string, the main question is, how is it converted to AuditLevel?

**Comment by pierre** [comment]
It's a string yet an enum for Swagger:

```

          {
            "name": "audit",
            "in": "query",
            "required": false,
            "type": "string",
            "default": "NONE",
            "enum": [
              "FULL",
              "MINIMAL",
              "NONE"
            ]
          }
```

The rest is done in our custom `KillbillJavaGenerator` (`extends AbstractJavaCodegen implements CodegenConfig`):

```
    @Override
    public String toEnumName(CodegenProperty property) {
        if (property.isEnum) {
            if (property._enum != null && property._enum.size() > 0) {
                final String guessedEnumType = findEnumType(property._enum.get(0));
                return guessedEnumType;
            } else if (property.datatypeWithEnum != null) {
                return property.datatypeWithEnum;
            }
        }
        System.err.println("Mising _enum for " + property.name);
        return sanitizeName(camelize(property.name));
    }

    private  String findEnumType(final String firstEnumValue) {
        for (final String key : apiEnums.keySet()) {
            Class claz = apiEnums.get(key);
            if (EnumSet.allOf(claz).iterator().next().toString().equals(firstEnumValue)) {
                return key;
            }

        }
        throw new IllegalStateException("Cannot find enum with first value " + firstEnumValue);
    }
```

**Comment by GregOriol** [comment]
@pierre This issue has been resolved with PR #39, I'm closing it (the rest of the discussion could become another PR later)
